### PR TITLE
✨Add automation examples for v1alpha2 clusters

### LINKED ISSUE #1018: Generate examples for v1alpha2
We need to be able to generate example clusters (to also include into release artifacts for each release). From v1alpha1 to v1alpha2, in v1alpha1 each provider vendored Cluster API's YAML and built on top of it. For v1alpha2 each provider is responsible for its own yaml because it does not vendor cluster-api.

Related: Examples generation scripts were removed here https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/994/files

/assign

### PR DESCRIPTION
**Pull request by vincepri**
Signed-off-by: Vince Prignano <vincepri@vmware.com>

**What this PR does / why we need it**:
This PR re-introduces the `examples/` folder, which gives users a way to play around with a simple cluster.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1018 

**Release note**:

```release-note

```

### CODE REVIEW

File: examples/addons.yaml [hunk 1]
```diff
@@ -0,0 +1,750 @@
+# Calico Version v3.6
```
Comment (detiber) [inline examples/addons.yaml]: Probably should update to 3.8 while we are adding manifests back: https://docs.projectcalico.org/v3.8/manifests/calico.yaml

File: examples/generate.sh [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+if [ $OVERWRITE -ne 1 ] && [ -d "$OUTPUT_DIR" ]; then
+  echo "ERR: Folder ${OUTPUT_DIR} already exists. Delete it manually before running this script."
+  exit 1
+fi
+
+mkdir -p "${OUTPUT_DIR}"
+
+# Generate cluster resources.
+kustomize build "${SOURCE_DIR}/cluster" | envsubst > "${CLUSTER_GENERATED_FILE}"
```
Comment (detiber) [inline examples/generate.sh]: should this use `kubectl kustomize` instead?

Comment (vincepri) [reply examples/generate.sh]: It fails for me

```
Error: rawResources failed to read Resources: Load from path ../crd failed: '../crd' must be a file (got d='/Users/vince/go/src/sigs.k8s.io/cluster-api-provider-aws/config/crd')
```

Comment (detiber) [reply examples/generate.sh]: Do we need to enforce a minimum version of kustomize then?

File: examples/addons.yaml (hunk 1 above)
Comment (vincepri) [reply examples/addons.yaml]: Updated 😊

File: examples/generate.sh (hunk 2 above)
Comment (vincepri) [reply examples/generate.sh]: Yeah, I've been using v3.1, I can add a check in this script

Comment (vincepri) [reply examples/generate.sh]: Added the apiVersion and Kind to the kustomize scripts, that should fix the version

File: examples/generate.sh [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+
+# Directories.
+SOURCE_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null && pwd )"
+OUTPUT_DIR=${OUTPUT_DIR:-${SOURCE_DIR}/_out}
+ENVSUBST=${ENVSUBST:-envsubst}
```
Comment (ncdc) [inline examples/generate.sh]: Add a check to make sure this exists/is in PATH?

File: examples/generate.sh [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+CLUSTERAWSADM=${CLUSTERAWSADM:-clusterawsadm}
```
Comment (ncdc) [inline examples/generate.sh]: Add a check to make sure this exists/is in PATH?

File: examples/generate.sh (hunk 2 above)
Comment (ncdc) [reply examples/generate.sh]: I'd rather use `kubectl kustomize` if possible. It sounds like we may need to be more creative with bases vs resources. (I am nowhere near a kustomize expert.)

File: examples/generate.sh [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+# Generate cluster resources.
+kustomize build "${SOURCE_DIR}/cluster" | envsubst > "${CLUSTER_GENERATED_FILE}"
+echo "Generated ${CLUSTER_GENERATED_FILE}"
+
+# Generate controlplane resources.
+kustomize build "${SOURCE_DIR}/controlplane" | envsubst > "${CONTROLPLANE_GENERATED_FILE}"
+echo "Generated ${CONTROLPLANE_GENERATED_FILE}"
+
+# Generate Cluster API provider components file.
+kustomize build "github.com/kubernetes-sigs/cluster-api//config/default/?ref=master" > "${COMPONENTS_CLUSTER_API_GENERATED_FILE}"
```
Comment (ncdc) [inline examples/generate.sh]: Is the extra slash in `cluster-api//config` a typo?

Also, I don't think `ref=master` is going to be viable in the long term.

File: examples/generate.sh [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+echo "Generated ${COMPONENTS_CLUSTER_API_GENERATED_FILE}"
+
+# Generate Kubeadm Bootstrap Provider components file.
+kustomize build "github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm//config/default/?ref=master" > "${COMPONENTS_KUBEADM_GENERATED_FILE}"
+echo "Generated ${COMPONENTS_KUBEADM_GENERATED_FILE}"
+
+# Generate AWS Credentials.
+AWS_B64ENCODED_CREDENTIALS="$(${CLUSTERAWSADM} alpha bootstrap encode-aws-credentials)"
```
Comment (ncdc) [inline examples/generate.sh]: This fails if `AWS_REGION` is unset. Should we add a check for this at the top of this script?

File: examples/generate.sh (hunk 2 above)
Comment (vincepri) [reply examples/generate.sh]: `kubectl kustomize` is a few versions behind and it's impossible right now to understand which version it's actually vendoring (at least I couldn't find it). For now, let's use plain `kustomize` and move to kubectl kustomize if/when it checks all the boxes.

File: examples/generate.sh (hunk 6 above)
Comment (vincepri) [reply examples/generate.sh]: `clusterawsadm` prints an error if it cannot generate find credentials, that should be enough

File: examples/generate.sh (hunk 5 above)
Comment (vincepri) [reply examples/generate.sh]: Not a typo 😄 It's to differentiate between the repository and the path

Comment (vincepri) [reply examples/generate.sh]: Once we release a version, we can reference the tag directly instead of `master`

File: examples/generate.sh [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+set -o errexit
+set -o nounset
+
+# Directories.
+SOURCE_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null && pwd )"
+OUTPUT_DIR=${OUTPUT_DIR:-${SOURCE_DIR}/_out}
+
+# Binaries
+ENVSUBST=${ENVSUBST:-envsubst}
+command -v "${ENVSUBST}" || echo -v "Cannot find ${ENVSUBST} in path."
```
Comment (ncdc) [inline examples/generate.sh]: redirect the output to `/dev/null`

File: examples/generate.sh [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+CLUSTERAWSADM=${CLUSTERAWSADM:-clusterawsadm}
+command -v "${CLUSTERAWSADM}" || echo -v "Cannot find ${CLUSTERAWSADM} in path, build it using 'make binaries' in this repository."
```
Comment (ncdc) [inline examples/generate.sh]: redirect to `/dev/null`

File: examples/generate.sh (hunk 2 above)
Comment (ncdc) [reply examples/generate.sh]: It's v2.0.3 in k/k master now (https://github.com/kubernetes/kubernetes/blob/d951d19ec4d8ec829e83d146200ea91b664b6fa6/go.mod#L175)

File: examples/generate.sh (hunk 6 above)
Comment (ncdc) [reply examples/generate.sh]: It took several seconds of running before it failed for me. I'll leave it up to you.

File: examples/generate.sh (hunk 2 above)
Comment (vincepri) [reply examples/generate.sh]: Yeah, we're using v3.1 within Cluster API repositories

File: examples/generate.sh (hunk 6 above)
Comment (vincepri) [reply examples/generate.sh]: Got it, I moved clusterawsadm as the first thing to generate. The main time is spent fetching the kustomize remote repos (CAPI, and kubeadm bootstrapper)

**Comment by vincepri** [comment]
@ncdc Good to go?

**Comment by ncdc** [comment]
Gah, needs rebase

**Comment by vincepri** [comment]
@ncdc rebased, thanks!

**Pull request merged**